source/common/io/io_uring_impl.h

```diff
@@ -29,13 +41,17 @@ class IoUringImpl : public IoUring, public ThreadLocal::ThreadLocalObject {
   IoUringResult prepareWritev(os_fd_t fd, const struct iovec* iovecs, unsigned nr_vecs,
                               off_t offset, void* user_data) override;
   IoUringResult prepareClose(os_fd_t fd, void* user_data) override;
+  IoUringResult prepareCancel(void* cancelling_user_data, void* user_data) override;
   IoUringResult submit() override;
+  void injectCompletion(os_fd_t fd, void* user_data, int32_t result) override;
+  void removeInjectedCompletion(os_fd_t fd) override;
 
 private:
   const uint32_t io_uring_size_;
   struct io_uring ring_ {};
   std::vector<struct io_uring_cqe*> cqes_;
   os_fd_t event_fd_{INVALID_SOCKET};
+  std::list<InjectedCompletion> injected_completions_;
```

Comment:
from usage I've seen so far, this doesn't need to be a doubly linked list, we could get away with forward_list.

What is expected operation workload? I'm expecting workload would be heavy on injectCompletion, forEveryCompletion more so than removal. Removal is O(n).

Reply:
Yes, it would be heavy on injectCompletion. The removal happened when socket closed. Also yes, forward_list is enough here.

Reply:
emm...the std::forward_list can't do the emplace_back. I still want the FIFO, I wasn't sure FILO works for us. And I suppose the injected_event list won't be huge.

Reply:
Shall we switch to `std::queue` if you want FIFO?